src/main/java/com/vinylteam/vinyl/util/impl/HardWaxParser.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    Set<String> getAllPagesByGenre(String genreLink) {
+        Set<String> allPages = getDocument(genreLink)
+                .stream()
+                .flatMap(document -> document.select(PRELIMINARY_PAGE_LINK_SELECTOR).stream())
+                .map(pageLink -> genreLink + URL_FILTER_PARAMETER + pageLink.attr("href").replace("?","&"))
+                .collect(toSet());
+        if (allPages.isEmpty()) {
+            return Set.of(genreLink);
+        }
+        return getAllPageLinksSet(allPages);
+    }
+
+    Set<String> getAllPageLinksSet(Set<String> pageLinks) {
+        int maxPageNumber = countPageLinks(pageLinks);
+        String pageLinkPattern = pageLinks.iterator().next();
+        log.debug("Pages found {'maxPageNumber':{}}", maxPageNumber);
+        Set<String> fullListOfPageLinks =
+                IntStream.rangeClosed(1, maxPageNumber)
+                        .mapToObj(pageNumber -> pageLinkPattern.replaceAll(PAGE_NUMBER_PATTERN.toString(), "page=" + pageNumber))
+                        .collect(toSet());
+        log.debug("Resulting set of page links (with no gaps in sequence) is {'pageLinks':{}}", pageLinks);
+        return fullListOfPageLinks;
+    }
+
+    int countPageLinks(Set<String> pageLinks) {
+        return pageLinks
+                .stream()
+                .map(PAGE_NUMBER_PATTERN::matcher)
+                .filter(Matcher::find)
+                .map(pageLinkMatcher -> pageLinkMatcher.group(1))
+                .map(Integer::parseInt)
+                .max(Comparator.naturalOrder())
+                .orElse(0);
+    }
+
+    Set<RawOffer> readOffersFromAllPages(Set<String> pageLinks) {
+        return pageLinks
+                .stream()
+                .flatMap(link -> readOffersFromOneLink(link).stream())
+                .collect(toSet());
+    }
+
+    Set<RawOffer> readOffersFromOneLink(String pageLink) {
+        String genre = getGenreFromLink(pageLink);
+        Set<RawOffer> offerLinks = this.getDocument(pageLink)
+                .stream()
+                .map(document -> document.select(OFFER_LIST_SELECTOR))
+                .flatMap(offersList -> offersList.select(ONE_VINYL_SELECTOR).stream())
+                .map(releaseElement -> getRawOfferFromElement(releaseElement, detailedParser))
+                .filter(this::isValid)
+                .collect(toSet());
+        offerLinks.forEach(offer -> offer.setGenre(genre));
+        log.info("Resulting set of offers for {'pageLink':{}} has size {'offerLinks.size':{}}", pageLink, offerLinks.size());
+        log.debug("Resulting set of offers for {'pageLink':{}} is {'offerLinks':{}}", pageLink, offerLinks);
+        return offerLinks;
+    }
+
+    String getGenreFromLink(String pageLink) {
+        Matcher matcher = GENRE_PATTERN.matcher(pageLink);
+        if (matcher.find()){
+            return matcher.group(1);
+        }
+        return "";
+    }
+
+    class HardWaxDetailedParserImpl extends DetailedVinylParserImpl {
+
```

Comment:
Why inner class? Shouldn't it be static?

Reply:
private?

Reply:
It is inner since this implementation is specific for hardwax.com parser and used only inside the parser.
it was supposed that an instance of the hardwax parser only will create instances of this inner class. The rest of parser tasks are covered by general implementation  - DetailedVinylParserImpl.
So the private modifier is a good idea, thanks!